src/test/scala/freeeventsourcing/ProcessTestSupport.scala

```diff
@@ -0,0 +1,138 @@
+      implicit
+      i: Inject[C#Error, E]
+    ) =
+      ExpectCommand(aggregateType, aggregate, command, Xor.left(i(result)))
+    def end = ExpectEnd
+  }
+
+  case class ProcessMatcher(expectations: List[Expectation], checkResult: Any ⇒ String Xor Unit) extends Matcher[M[_]] {
+    def apply(process: M[_]) = {
+      val x = process.foldMap(Implementation.Transform)
+      val y = x
+        .run(expectations)
+        .leftMap(error ⇒ s"did not match the expectations: ${error}")
+        .flatMap {
+          case (Nil, r)       ⇒ Xor.right(r)
+          case (remaining, _) ⇒ Xor.left(s"remaining expectations: ${remaining.mkString(", ")}")
+        }
+        .flatMap(checkResult)
+      new MatchResult(
+        y.isRight,
+        y.fold(identity, _ ⇒ "?"),
+        s"matched the process expectations but should not have"
+      )
+    }
+  }
+
+  type Action[+A] = ProcessAction[BC, A]
+  type M[A] = ProcessMonad[BC, A]
+
+  sealed trait Expectation
+  object Expectation {
+    case class ExpectAwaitEvent[S <: WithEventType](selector: S, result: S#Event) extends Expectation
+    case class ExpectWaitUntil(instant: Instant) extends Expectation
+    case class ExpectCommand[A <: Aggregate, Cmd <: A#Command](
+      aggregateType: A, aggregate: A#Id, command: A#Command, result: Cmd#Error Xor Unit
+    ) extends Expectation
+    case object ExpectEnd extends Expectation
+  }
+
+  private[this] object Implementation {
+    import Expectation._
+
+    type OrFail[A] = Xor[String, A]
+    type Expectations[A] = StateT[OrFail, List[Expectation], A]
+    object Expectations {
+      def next(received: String): Expectations[Expectation] = lift {
+        case h :: t ⇒ Xor.right((t, h))
+        case Nil    ⇒ Xor.left(s"No more expectations, but got ${received}")
+      }
+
+      def check[A](action: Action[A]): Expectations[A] = next(action.toString).flatMap { exp ⇒
+        compare.lift((action, exp)).getOrElse {
+          lifted(Xor.left(s"Expected ${exp} but got ${action}"))
+        }
+      }
+
+      private[this] def compare[A]: PartialFunction[(Action[A], Expectation), Expectations[A]] = {
+        case (AwaitEvent(s1), ExpectAwaitEvent(s2, result)) ⇒ lifted {
+          if (s1 == s2) Xor.right(result)
+          else Xor.left(s"AwaitEvent: selector mismatch ${s1} != ${s2}")
+        }
+
+        case (WaitUntil(i1), ExpectWaitUntil(i2)) ⇒ lifted {
+          if (i1 == i2) Xor.right(())
+          else Xor.left(s"WaitUntil: time is different: ${i1} != ${i2}")
+        }
+
+        case (End(), ExpectEnd) ⇒ lifted {
+          Xor.right(())
+        }
+
+        case (ex @ Execute(at1, a1, c1, errorHandler), e @ ExpectCommand(at2, a2, c2, r)) ⇒
+          if (at1 != at2) lifted(Xor.left(s"Execute: Aggregate type is different: ${at1} != ${at2}"))
+          else if (a1 != a2) lifted(Xor.left(s"Execute: Aggregate id is different: ${a1} != ${a2}"))
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Consider using case matching instead of else if blocks](https://www.codacy.com/app/msiegenthaler/free-event-sourcing/file/2580036554/issues/source?bid=3298199&fileBranchId=3298202#l112)